buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java

```diff
@@ -0,0 +1,331 @@
+ */
+package org.elasticsearch.gradle;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.unmodifiableList;
+
+/**
+ * Parse the Java source file containing the versions declarations and use the known rules to figure out which are all
+ * the version the current one is wire and index compatible with.
+ * On top of this, figure out which of these are unreleased and provide the branch they can be built from.
+ * <p>
+ * Note that in this context, currentVersion is the unreleased version this build operates on.
+ * At any point in time there will surely be four such unreleased versions being worked on,
+ * thus currentVersion will be one of these:
+ * - the unreleased <b>major</b>, a+1.0.0 on the `master` branch
+ * - the unreleased <b>minor</b>,  a.b.0 ( b != 0) on the `a.x` branch
+ * - the unreleased <b>maintenance</b>, a.b.c (c != 0) on the `a.b` branch
+ * - the unreleased <b>bugfix</b>, a-1.d.e ( d != 0, e != 0) on the `(a-1).d` branch
+ * In addition to these, there will be a fifth one when a minor reaches feature freeze, we call this the <i>staged</i>
+ * version:
+ * - the unreleased <b>staged</b>, a.b-2.0 (b &gt; 2) on the `a.(b-2)` branch
+ * Each build is only concerned with possible unreleased versions before it, as those are the ones that need to be tested
+ * for backwards compatibility. We never look forward, and don't add forward facing version number to branches of previous
+ * version.
+ * <p>
+ * The build know the current version, and we parse server code to find the rest making sure that these match.
+ * We can reliably figure out which the unreleased versions are due to the convention of always adding the next unreleased
+ * version number to server in all branches when a version is released.
+ * This convention is enforced by checking the versions we consider to be unreleased against an
+ * authoritative source (maven central).
+ */
+public class VersionCollection {
+
+    private static final Pattern LINE_PATTERN = Pattern.compile(
+        "\\W+public static final Version V_(\\d+)_(\\d+)_(\\d+)(_alpha\\d+|_beta\\d+|_rc\\d+)? .*"
+    );
+
+    private final Version currentVersion;
+    private final Map<Integer, List<Version>> groupByMajor;
+
+    public class UnreleasedVersionDescription {
+        private final Version version;
+        private final String branch;
+        private final String gradleProjectName;
+
+        UnreleasedVersionDescription(Version version, String branch, String gradleProjectName) {
+            this.version = version;
+            this.branch = branch;
+            this.gradleProjectName = gradleProjectName;
+        }
+
+        public Version getVersion() {
```

Comment:
Do we really need getters when the members are all final? I think they call all be `public final`?